scalding-core/src/test/scala/com/twitter/scalding/CoreTest.scala

```diff
@@ -290,6 +290,35 @@ class JoinTest extends WordSpec with Matchers {
       .run
       .finish()
   }
+
+  "join" should {
+    "correctly work with fold" in {
+      case class Value(opt: Option[Int] = None)
+      case class TotalValue(k: String, opt: Option[Int])
+
+      val left =
+        TypedPipe.from(List("a" -> 1, "b" -> 2))
+          .groupBy(_._1)
+          .foldLeft(Value()) {
+            case (acc, (_, int)) => acc.copy(Some(int))
+          }
+
+      val right =
+        TypedPipe.from(List("c" -> 3, "b" -> 2))
+
+
+      val res = left.join(right).map {
```

Comment:
also interesting if this manifests only on the left. A join should be symmetric: we can make a law to test that `a.join(b).values == b.join(a).values.swap` I wonder if that law is being violated somehow.

We have random tests in OptimizationTest that we could add this to.

Reply:
also, is the .map needed to trigger it?

Reply:
no, the map is not necessary to trigger.